add map between severity and overdue minutes

HNOONa-0 opened the pull request:
I have added an `unordered_map` that can store a severity and it's corresponding overdue time to the `LogCleaner` object.  The modifications should be backwards compatible. If this idea is correct, we can preform some optimizations, like keeping a cache of filepaths whose severities have been computed.

Fixes: #882

sergiud commented on src/logging.cc:
```diff
@@ -337,6 +338,16 @@ const char* GetLogSeverityName(LogSeverity severity) {
 
+LogSeverity FindFilepathLogSeverity(const std::string& filepath) {
+  for (int i = GLOG_INFO; i < NUM_SEVERITIES; i++) {
+    if (filepath.find(GetLogSeverityName(static_cast<LogSeverity>(i))) != std::string::npos) {
```
```suggestion
    if (filepath.rfind(GetLogSeverityName(static_cast<LogSeverity>(i))) != std::string::npos) {
```
Search from right since the severity is used as a suffix.

sergiud commented on src/logging.cc:
```diff
@@ -337,6 +338,16 @@ const char* GetLogSeverityName(LogSeverity severity) {
+      return static_cast<LogSeverity>(i);
+    }
+  }
+  // assume that the file has the INFO severity
```
Do not assume anything and simply return a not found sentinel.

sergiud commented on src/logging.cc:
```diff
@@ -1286,7 +1299,14 @@ LogCleaner::LogCleaner() = default;
   enabled_ = true;
-  overdue_ = overdue;
+  for (int i = GLOG_INFO; i < NUM_SEVERITIES; i++) {
+    overdue_per_severity_[static_cast<LogSeverity>(i)] = overdue;
```
Fill the map only with the severity set by the user.

sergiud commented on src/logging.cc:
```diff
@@ -2627,6 +2654,10 @@ void EnableLogCleaner(const std::chrono::minutes& overdue) {
   log_cleaner.Enable(overdue);
 }
 
+void EnableLogCleaner(std::unordered_map<LogSeverity, std::chrono::minutes>& overdue_per_severity) {
```
Exposing the map as a parameter makes the calls verbose and the API usage unnecessarily complicated. Instead use the following signature:
```suggestion
void EnableLogCleaner(LogSeverity severity, const std::chrono::minutes& overdue) {
```

sergiud commented on src/logging.cc:
```diff
@@ -2627,6 +2654,10 @@ void EnableLogCleaner(const std::chrono::minutes& overdue) {
+  log_cleaner.Enable(overdue_per_severity);
+}
+
 void DisableLogCleaner() { log_cleaner.Disable(); }
```
For symmetry, there should be an API call for disabling the log cleaner for specific severity.

sergiud reviewed (changes requested):
Thanks for the PR and apologies for the late reply.

The changes are going into the right direction. Please add some unit tests. For this, you can extend the existing log cleaner tests.

HNOONa-0 commented:
Hey @sergiud , I have made some improvements, I switched from a map to a vector for better lookup speed and cleaner implementation. I also addressed all the issues you pointed out. PTAL!

HNOONa-0 commented on src/logging.cc:
```diff
@@ -459,9 +471,7 @@ class LogCleaner {
-  bool enabled_{false};
-  std::chrono::minutes overdue_{
-      std::chrono::duration<int, std::ratio<kSecondsInWeek>>{1}};
+  vector<std::chrono::minutes> overdue_per_severity_;
```
Memory is so small, its the same size of `NUM_SEVERITIES`, which is 4 and the size of the vector is static.

I thought this would be more efficient and simpler but I can make those changes.

sergiud replied:
You are thinking about desktop systems where you have plenty of memory to spare. glog, however, is also used on embedded hardware where you cannot afford allocating memory and leave it unused.

Instead of an `std::vector` you could also use `std::array`. This would avoid heap allocation.

HNOONa-0 commented:
Hey @sergiud, I initially made the wrong assumptions, I thought log files must have severity in their names, but that’s not the case. I have added methods to enable, disable logging for both types of files (with and without severity in their names) and exported those functions. I also added a test file that checks for deferred cleanup logic, but I haven’t integrated it with CMake yet.
